[#1705] Variables documentation (#1893)
open-formulieren/open-forms: docs/developers/backend/core/variables.rst

Variables
=========

There are two models for variables, ``FormVariable`` and ``SubmissionValueVariable``. ``FormVariable`` are related to
a form while ``SubmissionValueVariable`` are related to a submission.

Form Variables
==============

A ``FormVariable`` can have one of two sources:

* **Component**: the variable corresponds to a component in the form.
* **User defined**:
  the variable is added by the person designing the form and doesn't explicitly appear in the form. It
  can however be used for calculations or in form fields. They can also be prefilled in the same way as form components.

The API for the ``FormVariables`` has a bulk create / update endpoint. This means that a ``PUT`` call to this endpoint
will delete all existing ``FormVariables`` related to a form and replace them with the data sent in the PUT call.
If there are any existing completed submissions, the ``SubmissionValueVariables`` will not be deleted, but they will no
longer be related to a ``FormVariable``. Any in progress submission will also have ``SubmissionValueVariables`` without

> a related ``FormVariable``. The result of this situation is that any data input by the user will not be saved.

Reviewer: that last sentence confuses me - if the form definition changes while someone is filling it out, it'll still persist the values for the now-configured form variables/components, right? (assuming the form keys didn't change, and if they did, the user can go back a step and fix the input)

Author: Now that I stepped through it I actually think there is a bug :thinking: The current behaviour is what I described (if a user is filling out the form, but the definition is updated, then the data is not saved). It happens because no form variables are retrieved from this query:
https://github.com/open-formulieren/open-forms/blob/master/src/openforms/submissions/models/submission_value_variable.py#L100
(the `saved_submission_value_variables.values_list( "form_variable__id", flat=True)` contains a list with `None`). So I need to look into it :thinking:

Author: https://github.com/open-formulieren/open-forms/issues/1928

.. note::

>    In Open Forms versions < 1.2, form components could have the same key across different steps. Now, there is a unique

Reviewer: we're 90% sure going to release open forms 2.0 rather than 1.2

   together constraint on the ``form`` and ``key`` attributes of the ``FormVariable``. So all components must have
   different keys *across the form*. Before upgrading to 1.2, it is important to fix any duplicate keys.

   There is the management command ``check_duplicate_component_keys`` to check for duplicate keys on an environment.

Static variables
----------------

> There is also the concept of static variables. These are not saved in the database and only live in memory.

Reviewer: ```suggestion
Static variables are a third type of (form) variables. These are not saved in the database and only live in memory.
```

* **User defined**:
  the variable is added by the person designing the form and doesn't explicitly appear in the form. It
  can however be used for calculations or in form fields. They can also be prefilled in the same way as form components.

The API for the ``FormVariables`` has a bulk create / update endpoint. This means that a ``PUT`` call to this endpoint
will delete all existing ``FormVariables`` related to a form and replace them with the data sent in the PUT call.
If there are any existing completed submissions, the ``SubmissionValueVariables`` will not be deleted, but they will no
longer be related to a ``FormVariable``. Any in progress submission will also have ``SubmissionValueVariables`` without
a related ``FormVariable``. The result of this situation is that any data input by the user will not be saved.

.. note::

   In Open Forms versions < 2.0, form components could have the same key across different steps. Now, there is a unique
   together constraint on the ``form`` and ``key`` attributes of the ``FormVariable``. So all components must have
   different keys *across the form*. Before upgrading to 2.0, it is important to fix any duplicate keys.

>    There is the management command ``check_duplicate_component_keys`` to check for duplicate keys on an environment.

Reviewer: ```suggestion
   Use the management command ``check_duplicate_component_keys`` to check for duplicate keys on an environment.
```

Static variables
----------------

> Static variables are a third type of (form). These are not saved in the database and only live in memory.

Reviewer: "third type of (form)"  looks like a mistake

        are initialised with the specified initial value and persisted to the database.

   * ``GET to /submissions/<submission_uuid>/steps/<submission_step_uuid>``:
     The ``SubmissionStepSerializer`` evaluates the form logic to dynamically update the form configuration. This loads
     the ``SubmissionValueVariablesState`` which contains the value of the variables and of the static data. When the
     logic updates this state, the changes are kept in memory and are not persisted to the database.

#. During logic evaluation:

   * ``POST /submissions/<submission_uuid>/steps/<submission_step_uuid>/_check_logic``:
     The endpoint receives any data input by the user in a particular step. This data is merged with data already
     present for any other step of the form and it is used to evaluate the form logic and update dynamically the form
     configuration.

#. Going to the next step (persisting a step to the database):

   * ``PUT /submissions/<submission_uuid>/steps/<submission_step_uuid>``:
     When the ``SubmissionStepSerializer`` is saved during this request, any ``SubmissionValueVariables`` related to it

>      is persisted to the database.

Reviewer: ```suggestion
     are persisted to the database.
```

>      After the serializer is saved, and ``SubmissionValueVariables`` unrelated to a particular step is persisted if its

Reviewer: ```suggestion
     After the serializer is saved, any ``SubmissionValueVariables`` unrelated to a particular step are persisted if its
```
?

Author: I went for 
```
After the serializer is saved, any ``SubmissionValueVariable`` unrelated to a particular step is persisted if its ..
```
